atc/metric/emitter/newrelic.go

```diff
@@ -1,41 +1,74 @@
 package emitter
 
 import (
+	"bufio"
 	"bytes"
+	"compress/gzip"
 	"encoding/json"
 	"fmt"
+	"github.com/concourse/flag"
+	"io"
+	"io/ioutil"
 	"net/http"
 	"strings"
+	"sync"
 	"time"
 
 	"code.cloudfoundry.org/lager"
 	"github.com/concourse/concourse/atc/metric"
 	"github.com/pkg/errors"
 )
 
+const (
+	NEWRELIC_PAYLOAD_MAX_SIZE            = 1024 * 1024
+	NEWRELIC_FLUSH_CHECK_INTERVAL_IN_SEC = 30
+)
+
 type (
 	stats struct {
 		created interface{}
 		deleted interface{}
 	}
 
 	NewRelicEmitter struct {
-		client     *http.Client
-		url        string
-		apikey     string
-		prefix     string
-		containers *stats
-		volumes    *stats
+		prefix       string
+		containers   *stats
+		volumes      *stats
+		batchEmitter *batchEmitter
 	}
 
 	NewRelicConfig struct {
-		AccountID     string `long:"newrelic-account-id" description:"New Relic Account ID"`
-		APIKey        string `long:"newrelic-api-key" description:"New Relic Insights API Key"`
-		ServicePrefix string `long:"newrelic-service-prefix" default:"" description:"An optional prefix for emitted New Relic events"`
+		AccountID          string `long:"newrelic-account-id" description:"New Relic Account ID"`
+		APIKey             string `long:"newrelic-api-key" description:"New Relic Insights API Key"`
+		ServicePrefix      string `long:"newrelic-service-prefix" default:"" description:"An optional prefix for emitted New Relic events"`
+		CompressionEnabled bool   `long:"newrelic-metric-compression" description:"New Relic payload compression flag"`
+		FlushInterval      int    `long:"newrelic-flush-interval" default:"60" description:"New Relic metric flush interval in seconds"`
```

Comment:
nit: we typically use [`time.Duration`](https://golang.org/pkg/time/#Duration) for these things - e.g.: https://github.com/concourse/concourse/blob/7a6048b7d6a9124c891b58d9560a9af8be80edaf/atc/metric/emitter/influxdb.go#L31

Reply:
Good suggestion. Recently I just learned, a redefined type like `time.Duration`, we could leverage its original type to do related operations(like `+`, `*`).